velox/expression/tests/SimpleFunctionTest.cpp

```diff
@@ -551,6 +552,109 @@ TEST_F(SimpleFunctionTest, nonDefaultNullBehavior) {
   assertEqualVectors(expected, result);
 }
 
+// NullFree tests:
+
+// Test that function with default contains nulls behavior won't get called when
+// inputs are all null.
+template <typename T>
+struct DefaultContainsNullsBehaviorFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool callNullFree(
+      out_type<bool>& out,
+      const null_free_arg_type<Array<int32_t>>&) {
+    out = true;
+
+    return true;
+  }
+};
+
+TEST_F(SimpleFunctionTest, defaultContainsNullsBehavior) {
+  registerFunction<DefaultContainsNullsBehaviorFunction, bool, Array<int32_t>>(
+      {"default_contains_nulls_behavior"});
+
+  // Make a vector with some null arrays, and some arrays containing nulls.
+  auto arrayVector = makeVectorWithNullArrays<int32_t>(
+      {std::nullopt, {{1, std::nullopt}}, {{std::nullopt, 2}}, {{1, 2, 3}}});
+
+  // Check that default contains nulls behavior functions don't get called on a
+  // null input.
+  auto result = evaluate<SimpleVector<bool>>(
+      "default_contains_nulls_behavior(c0)", makeRowVector({arrayVector}));
+  auto expected = makeNullableFlatVector<bool>(
+      {std::nullopt, std::nullopt, std::nullopt, true});
+  assertEqualVectors(expected, result);
+}
+
+// Test that function with non-default contains nulls behavior.
+// Returns 0 for null and input[0] * 2 if call is invoked.
+// Returns input[0] * 8 if callNullFree is invoked.
+template <typename T>
+struct NonDefaultContainsNullsBehaviorFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<int32_t>& out,
+      const arg_type<Array<int32_t>>* input) {
+    if (input == nullptr) {
+      out = 0;
```

Comment:
nit: since we only want to test which function is called
maybe consider the body of the functions to be out =1 and out =2
or out = kCallNullable; 
and out = kCallNullFree

Reply:
I was trying to use this to also verify that the inputs were passed correctly (admittedly it's totally hacky), let me try to think of a better way to do this.